VestingSchedule.sol

```diff
@@ -0,0 +1,227 @@
+     */
+    function duration() external view returns (uint256) {
+        return _duration;
+    }
+
+    /**
+     * @return true if the vesting is revocable.
+     */
+    function revocable() external view returns (bool) {
+        return _revocable;
+    }
+
+    /**
+     * @return the cummulative amount of tokens released.
+     */
+    function released() external view returns (uint256) {
+        return _released;
+    }
+
+    /**
+     * @return true if the vesting is revoked.
+     */
+    function revoked(address token) external view returns (bool) {
+        return _revoked;
+    }
+
+    /**
+     * @return the timestamp of the revocation.
+     */
+    function revokeTime(address token) external view returns (bool) {
+        return _revokeTime;
+    }
+
+    /**
+     * @notice Transfers available released tokens from the vesting back to beneficiary.
+     */
+    function withdraw() external onlyBeneficiary {
+        uint256 unreleased = _releasableAmount(block.timestamp);
+
+        require(unreleased > 0, "No unreleased tokens are due for withdraw");
+
+        _released = _released.add(unreleased);
+
+        token.safeTransfer(_beneficiary, unreleased);
+
+        emit VestingWithdrawn(msg.sender, unreleased, block.timestamp);
+    }
+
+    /**
+     * @notice Allows only the revoker to revoke the vesting. Tokens already vested
+     * remain in the contract, the rest are returned to the _refundDestination.
+     * @param revokeTime the revocation timestamp
+     * @dev revokeTime the revocation timestamp. If is less than the current block timestamp, it is set equal
+     */
+    function revoke(revokeTime) external onlyRevoker {
+        require(_revocable, "Revoking is not allowed");
+        require(!_revoked, "Vesting already revoked");
+
+        uint256 revokeTimestamp = revokeTime > block.timestamp ? revokeTime : block.timestamp;
+
+        uint256 balance = getGoldToken().balanceOf(address(this));
+        uint256 unreleased = _releasableAmount(revokeTimestamp);
+        uint256 refund = balance.sub(unreleased);
+
+        _revoked = true;
+        _revokeTime = revokeTimestamp;
+
+        getGoldToken().transfer(_refundDestination, refund);
+
+        emit VestingRevoked(msg.sender, _refundDestination, refund, _revokeTime);
+    }
+
+    /**
+     * @dev Calculates the amount that has already vested but hasn't been withdrawn (released) yet.
+     * @param timestamp the timestamp at which the calculate the releasable amount
+     */
+    function _releasableAmount(uint256 timestamp) private view returns (uint256) {
+        return _vestedAmount(timestamp).sub(_released);
+    }
+
+    /**
+     * @dev Calculates the amount that has already vested.
+     * @param timestamp the timestamp at which the calculate the already vested amount
+     */
+    function _vestedAmount(uint256 timestamp) private view returns (uint256) {
+        uint256 currentBalance = getGoldToken().balanceOf(address(this));
```

Comment:
I am not sure you should base your calculation based on the balanceOf(this) as the vesting contract might transfer these tokens to the LockedGold contract. Arent you looking for a logic based around vestedAmount?

Reply:
covered in the new revision